File: fsharp/rna-transcription/RnaTranscription.fs
module RnaTranscription

    open System

    let private transciber (nucleotide:char):char =
        match nucleotide with
        | 'G' -> 'C'
        | 'C' -> 'G'
        | 'T' -> 'A'
        | 'A' -> 'U'
        | _ -> failwith "Invalid Nucleotide"

    let toRna (dna: string):string =
        dna.ToCharArray()
        |> Array.map transciber
        |> String

File: solutions/scala/rna-transcription/1/src/main/scala/RnaTranscription.scala
object RnaTranscription
{
    private def bases:Array[Char] = Array('G', 'C', 'T', 'A')

    private def validate(dna: String):Option[String] =
    {
        dna.forall(bases.contains) match
        {
            case true => Some(dna)
            case false => None
        }
    }

    private def transciber(nucleotide:Char):Char =
    {
        nucleotide match
        {
            case 'G' => 'C'
            case 'C' => 'G'
            case 'T' => 'A'
            case 'A' => 'U'
        }
    }

    def toRna(dna: String):Option[String] = validate(dna).map(valid => valid.map(transciber))
}